migrator/DataImporter.java

```diff
@@ -34,286 +35,539 @@
+        }
+
+        protected void recordAttributeCreated(ByteArray iid, String originalID) {
+            idMapper.put(originalID, iid);
+        }
+
+        Thing getThing(String originalID) {
+            ByteArray newIID;
+            if ((newIID = originalIDsToBufferedIIDs.get(originalID)) == null && (newIID = idMapper.get(originalID)) == null) {
+                throw TypeDBException.of(ILLEGAL_STATE);
+            } else {
+                Thing thing = transaction.concepts().getThing(newIID);
+                assert thing != null;
+                return thing;
+            }
+        }
+
+        boolean isImported(String originalID) {
+            return originalIDsToBufferedIIDs.containsKey(originalID) || idMapper.contains(originalID);
+        }
+
+        int insertOwnerships(String originalId, List<DataProto.Item.OwnedAttribute> ownedMsgs) {
+            Thing owner = getThing(originalId);
+            int ownerships = 0;
+            for (DataProto.Item.OwnedAttribute ownedMsg : ownedMsgs) {
+                Thing attrThing = getThing(ownedMsg.getId());
+                assert attrThing != null;
+                owner.setHas(attrThing.asAttribute());
+                ownerships++;
+            }
+            status.ownershipCount.addAndGet(ownerships);
+            return ownerships;
+        }
+
+        private void commitBatch() {
+            transaction.commit();
+            ((RocksTransaction.Data) transaction).bufferedToPersistedThingIIDs().forEachRemaining(pair -> {
+                idMapper.put(bufferedIIDsToOriginalIds.get(pair.first()), pair.second());
+            });
+            bufferedIIDsToOriginalIds.clear();
+            originalIDsToBufferedIIDs.clear();
         }
     }
 
-    private void insertAttribute(DataProto.Item.Attribute attributeMsg) {
-        AttributeType attributeType = tx.concepts().getAttributeType(relabel(attributeMsg.getLabel()));
-        if (attributeType != null) {
-            DataProto.ValueObject valueMsg = attributeMsg.getValue();
-            Attribute attribute;
-            switch (valueMsg.getValueCase()) {
-                case STRING:
-                    attribute = attributeType.asString().put(valueMsg.getString());
-                    break;
-                case BOOLEAN:
-                    attribute = attributeType.asBoolean().put(valueMsg.getBoolean());
-                    break;
-                case LONG:
-                    attribute = attributeType.asLong().put(valueMsg.getLong());
-                    break;
-                case DOUBLE:
-                    attribute = attributeType.asDouble().put(valueMsg.getDouble());
-                    break;
-                case DATETIME:
-                    attribute = attributeType.asDateTime().put(
-                            Instant.ofEpochMilli(valueMsg.getDatetime()).atZone(ZoneId.of("Z")).toLocalDateTime());
-                    break;
+    private class InitAndAttributes extends ImportWorker {
```

Comment:
in the first pass of the file we load attributes, and search for the checksum at the end of the file